### PR TITLE
ISSUE-5934: Support read/write properties from/to Message in flink pulsar consumer/producer

### LINKED ISSUE #5934: Can support read/write properties from/to Message in flink pulsar consumer/producer
**Is your feature request related to a problem? Please describe.**
Currently, flink pulsar consumer/producer can not read/write properties from/to Message, we hope to support it.

**Describe the solution you'd like**
A) For consumer, 
change method `deserialize(Message message)’` access level from private to proteced in class 'PulsarConsumerSource', and we can overrided it in derived class.
B) For producer,
1. add a method like following code,  and we can overrided it in derived class.
`protected Map<String, String> generateProperties(IN value) {
        return new HashMap<>();
 }`
2. invoke TypedMessageBuilder.properties() method addtional in `invoke(IN value, Context context)` method at class 'FlinkPulsarProducer'.

### PR DESCRIPTION
**Pull request by duli559**
Fix #5934

Motivation
Support read/write properties from/to Message in flink pulsar consumer/producer, and you can override it in your derived class

Modifications

1. modify `PulsarConsumerSource.deserialize` access right from 'private' to 'protected'.
2. add method `protected Map<String, String> generateProperties(T value)` in class `FlinkPulsarProducer`, and invoked in `TypedMessageBuilder.properties()` to add it in pulsar Message.

### CODE REVIEW

**Comment by sijie** [comment]
@yjshen can you please help review this pull request?

**Comment by yjshen** [comment]
LGTM.

**Comment by sijie** [comment]
@yjshen how does this user provide the properties using this code?

**Comment by yjshen** [comment]
As discussed in #5934, users need to subclass `FlinkPulsarProducer`, overwrite `generateProperties` proposed by this PR, and generate properties based on:
- the current value to write
- constant provided in subclass
- `RuntimeContext` from the base `RichSinkFunction`.

**Comment by sijie** [comment]
@yjshen sorry that I missed the discussion. I think it is better to pass in a function `PropertitiesExtractor` like KeyExtractor rather than inheriting FlinkPulsarProducer, no?

**Comment by yjshen** [comment]
@sijie I agree. 
@duli559 could you please check sijie's advice above?

**Comment by duli559** [comment]
@sijie @yjshen , it's a good advice, i change it.

File: pulsar-flink/src/main/java/org/apache/flink/streaming/connectors/pulsar/partitioner/PulsarPropertiesExtractor.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Extract message properties from a value or others.
+ */
+public interface PulsarPropertiesExtractor<T> extends Serializable {
+
+    PulsarPropertiesExtractor EMPTY = in -> new HashMap<>();
```
Comment (sijie) [inline pulsar-flink/src/main/java/org/apache/flink/streaming/connectors/pulsar/partitioner/PulsarPropertiesExtractor.java]: I don't think it is a good idea to create a new hash map for every event. It is better to return `Collections.emptyMap`.

Comment (duli559) [reply pulsar-flink/src/main/java/org/apache/flink/streaming/connectors/pulsar/partitioner/PulsarPropertiesExtractor.java]: @sijie ,Thanks for your review, it is really more efficient which change it to a reusable object.

**Comment by duli559** [comment]
why this ci build is failed? I feel a little strange, some other pull requests have the same problem, @yjshen

**Comment by tuteng** [comment]
retest this please

**Comment by sijie** [comment]
> why this ci build is failed? I feel a little strange, some other pull requests have the same problem,

@duli559 the current Jenkins is a bit flaky. We are working on migrating to Github Actions.

**Comment by duli559** [comment]
> > why this ci build is failed? I feel a little strange, some other pull requests have the same problem,
> 
> @duli559 the current Jenkins is a bit flaky. We are working on migrating to Github Actions.

@sijie , it's great, and i have changed my commited code with your earlier review advice.

**Comment by duli559** [comment]
@yjshen, it seems some instable unit tests failed

**Comment by yjshen** [comment]
@duli559 could you please merge the master? (git fetch apache master && git merge apache/master)

**Pull request merged**